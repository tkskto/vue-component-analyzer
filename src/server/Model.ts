import {FORMAT} from './Constant';

class Model {
  private _resourceRoot = '.';

  private _format = FORMAT.BROWSER;

  private _outDirectory = 'out';

  private _port = '8888';

  private _isSilentMode = false;

  get resourceRoot(): string {
    return this._resourceRoot;
  }

  set resourceRoot(value: string) {
    this._resourceRoot = value;
  }

  get format(): string {
    return this._format;
  }

  set format(value: string) {
    this._format = value;
  }

  get outDirectory(): string {
    return this._outDirectory;
  }

  set outDirectory(value: string) {
    this._outDirectory = value;
  }

  get port(): string {
    return this._port;
  }

  set port(value: string) {
    this._port = value;
  }

  get isSilentMode(): boolean {
    return this._isSilentMode;
  }

  set isSilentMode(value: boolean) {
    this._isSilentMode = value;
  }
}

export const model = new Model();
