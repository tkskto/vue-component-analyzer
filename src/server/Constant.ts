export const FORMAT = {
  BROWSER: 'browser',
  JSON: 'json',
  BOTH: 'both',
  HTML: 'html',
};
