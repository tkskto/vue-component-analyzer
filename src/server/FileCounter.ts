const cwd = process.cwd();

export class FileCounter {
  private _count: {[key: string]: number} = {};

  public add(_filename: string): void {
    const shortFilename = _filename.replace(cwd, '');

    if (Object.prototype.hasOwnProperty.call(this._count, shortFilename)) {
      this._count[shortFilename]++;
    } else {
      this._count[shortFilename] = 1;
    }
  }

  public clear(): void {
    this._count = {};
  }

  public get result(): {[key: string]: number} {
    return this._count;
  }
}

export const fileCounter = new FileCounter();
