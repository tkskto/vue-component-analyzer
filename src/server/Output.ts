import {AnalyzeReport} from '../../types';
import {Renderer} from '../client/renderer';
import {Model} from '../client/model';
import {mkdirp} from 'mkdirp';
import path from 'path';
import {writeFileSync} from 'fs';
import {render} from 'ejs';
import {fileURLToPath} from 'url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);
const projectRoot = path.resolve(__dirname, '..');

export async function writeFileExtra(_filename: string, data: string) {
  const filename = _filename.endsWith('.json') ? _filename : `${_filename}/result.json`;

  try {
    await mkdirp(path.dirname(filename));

    writeFileSync(filename, data);
    console.log(`output completed ${filename}`);
  } catch (err: unknown) {
    if (err instanceof Error) {
      throw new Error(err.message);
    }
  }
}

export async function writeHTML(_filename: string, result: AnalyzeReport) {
  const filename = _filename.endsWith('.html') ? _filename : `${_filename}/result.html`;

  try {
    await mkdirp(path.dirname(filename));

    const clientModel = new Model();

    clientModel.data = result;
    clientModel.env = 'node';

    const renderer = new Renderer(clientModel);
    const htmlString = renderer.render();
    const html = render('<%- include("/views/static.ejs", {html: html}); %>', {html: htmlString}, {root: `${projectRoot}`});

    writeFileSync(filename, html);
    console.log(`output completed ${filename}`);
  } catch (err: unknown) {
    if (err instanceof Error) {
      throw new Error(err.message);
    }
  }
}
