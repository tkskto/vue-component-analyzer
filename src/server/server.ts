// for execute from npm scripts or commandline.

import AnalyzeReport = vueComponentAnalyzer.AnalyzeReport;
import express from 'express';
import {WebSocketServer} from 'ws';
import path from 'path';
import http from 'http';
import {renderFile} from 'ejs';
import opener from 'opener';
import {fileURLToPath} from 'url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);
const projectRoot = path.resolve(__dirname, '..');

/**
 * start web socket server and connect when finished generate component tree.
 */
export const startServer = (port: string, json: AnalyzeReport): void => {
  const HOST = '127.0.0.1';

  // establish local web server.
  const app = express();

  // use EJS to pass variables.
  app.engine('ejs', renderFile);
  app.set('view engine', 'ejs');
  app.set('views', `${projectRoot}/views`);
  app.use(express.static(`${projectRoot}/`));

  // use same server for WebSocket Server.
  const server = http.createServer(app);
  const wss = new WebSocketServer({
    server,
  });

  app.use('/', (req: express.Request, res: express.Response) => {
    res.render('viewer', {
      mode: 'server',
      title: 'analyze report',
      enableWebSocket: true,
    });
  });

  server.on('error', (err: Error) => {
    console.log(err);

    if (server.listening) {
      server.close((serverErr: Error | undefined) => {
        console.log(serverErr);
      });
    }
  });

  server.listen(Number(port), HOST, () => {
    const address = server?.address();
    let addressPort = port;

    if (typeof address === 'string') {
      addressPort = address;
    } else if (typeof address === 'object') {
      addressPort = String(address?.port) || port;
    }

    const url = `http://${HOST}:${addressPort}/`;

    console.log(`Vue Component Analyzer is started at ${url}`);
    console.log('Use \'Ctrl+C\' to close it');

    opener(url);

    wss.on('connection', (ws: WebSocket) => {
      // pass tree data to browser.
      ws.send(JSON.stringify(json));

      ws.addEventListener('error', () => {
        console.error('Something went to wrong on web socket.');
      });
    });
  });
};
