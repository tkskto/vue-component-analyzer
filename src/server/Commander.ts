import {FORMAT} from './Constant';
import {program} from 'commander';

export const getOptions = (_argv: string[]): {
  dir: string,
  format: string,
  out: string,
  port: string,
  silent: boolean,
} => {
  program.option('--dir [dir]', 'root directory of src.', 'src');
  program.option('-f, --format [type]', 'Add the specified type of report [browser, json or both]', FORMAT.BROWSER);
  program.option('-o, --out [dir]', 'output directory (enable with setting --format option to "json" or "both")', 'out');
  program.option('-p, --port [number]', 'select a port number for the local server', '8888');
  program.option('--silent', 'do not show logs with silent flag');
  program.parse(_argv);

  return program.opts();
};
