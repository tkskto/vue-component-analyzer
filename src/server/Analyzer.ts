import FileReport = vueComponentAnalyzer.FileReport;
import {resolveFile} from './utils';
import {readFileSync, statSync} from 'fs';
import {resolve, extname} from 'path';
import {FileCounter} from './FileCounter';
import {VueComponent} from './VueComponent';
import {model} from './Model';
const cwd = process.cwd();

class Analyzer {
  private readonly _counter: FileCounter;

  constructor() {
    this._counter = new FileCounter();
  }

  /**
   * get import tree.
   * @param {string} fileName entry file name
   * @param {string[]} parents entries of file name
   * @param {boolean} isTest whether it through the test. if this doesn't exist lastModifiedTime will always fail on the GitHub Actions.
   */
  public getImportDeclarationTree = (fileName: string, parents: string[], isTest = false): FileReport => {
    const filename = resolve(cwd, fileName);
    // get filename without current working directory.
    const shortFilename = filename.replace(cwd, '');
    // get file statistic
    const stat = statSync(filename);
    // make a filename list for detecting circular dependency.
    const ancestorList: string[] = parents.concat();

    ancestorList.push(fileName);

    if (!model.isSilentMode) {
      console.log(`read ${filename}`);
    }

    // increment count of this file.
    this._counter.add(shortFilename);

    // if this file is not Vue Component file, return only filename and stat.
    if (extname(filename) === '' || extname(filename) !== '.vue') {
      return {
        name: shortFilename,
        props: '',
        size: stat.size,
        lastModifiedTime: isTest ? 0 : Number(stat.mtimeMs.toFixed(0)),
        children: [],
      };
    }

    const contents = readFileSync(filename, 'utf-8');
    const component = new VueComponent(shortFilename, contents, stat);

    try {
      // if we get, read imported file recursive.
      for (let i = 0, len = component.importDeclaration.length; i < len; i++) {
        // TODO: support other types? (value might be RegExp, or number, boolean.)
        const source = String(component.importDeclaration[i].source.value);

        if (source) {
          const nextFilename = resolveFile(source, filename);

          if (nextFilename) {
            if (parents.includes(nextFilename)) {
              console.warn(`Circular dependency detected between ${nextFilename} and ${filename}`);
            } else {
              component.addChildReport(this.getImportDeclarationTree(nextFilename, ancestorList, isTest));
            }
          }
        }
      }
    } catch (err) {
      console.error(`Something went wrong with reading ${filename}`);
      console.error(err.message);
    }

    return component.getFileReport(isTest);
  };

  get counter(): FileCounter {
    return this._counter;
  }
}

export const analyzer = new Analyzer();
