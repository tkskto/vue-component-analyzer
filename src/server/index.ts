// for execute from npm scripts or commandline.

import {FORMAT} from './Constant';
import {model} from './Model';
import {getOptions} from './Commander';
import {analyzer} from './Analyzer';
import {startServer} from './server';
import {writeFileSync} from 'fs';
import {globby} from 'globby';
import path from 'path';
import mkdirp from 'mkdirp';

// TODO: add log system
// TODO: add error system

function writeFileExtra(filename: string, data: string) {
  mkdirp(path.dirname(filename)).then(() => {
    writeFileSync(filename, data);
  }).catch((err: Error) => {
    if (err) {
      throw new Error(err.message);
    }
  });
}

export const vueComponentAnalyze = async () => {
  try {
    // get commander options.
    const argv = getOptions(process.argv);

    model.resourceRoot = argv.dir;
    model.format = argv.format;
    model.outDirectory = argv.out;
    model.port = argv.port;
    model.isSilentMode = argv.silent || false;

    if (argv.format !== FORMAT.BROWSER && argv.format !== FORMAT.JSON && argv.format !== FORMAT.BOTH) {
      console.error(`not support ${argv.format} format.`);
    }

    console.log('start analyzing.');

    // get vue entries from dir
    const entries = await globby([argv.dir, '!**/node_modules/**'], {
      expandDirectories: {
        extensions: ['vue'],
      },
    });

    if (entries.length === 0) {
      console.log('There is no entry file.');
    }

    const entriesData = [];

    for (let i = 0, len = entries.length; i < len; i++) {
      const entryFile = entries[i];
      const children = analyzer.getImportDeclarationTree(entryFile, []);

      entriesData.push(children);
    }

    const result = {
      entries: entriesData,
      count: analyzer.counter.count,
    };

    if (argv.format === FORMAT.BOTH) {
      startServer(argv.port, result);
      writeFileExtra(path.resolve(process.cwd(), `${argv.out}/result.json`), JSON.stringify(result, null, 4));
    } else if (argv.format === FORMAT.BROWSER) {
      startServer(argv.port, result);
    } else if (argv.format === FORMAT.JSON) {
      writeFileExtra(path.resolve(process.cwd(), `${argv.out}/result.json`), JSON.stringify(result, null, 4));
    }

    console.log('finished analyzing.');
  } catch (err) {
    if (err instanceof Error) {
      console.error(err.message);
    }
  }
};
