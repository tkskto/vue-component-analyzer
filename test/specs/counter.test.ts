import {join} from 'path';
import {analyzer} from '../../src/server/Analyzer';
const fixturesDir = join(__dirname, '../fixture/');

describe('counter\'s count test', () => {
  test('counterTest.vue', async () => {
    const filename = 'counterTest/counterTest';
    await analyzer.getImportDeclarationTree(join(fixturesDir, `${filename}.vue`), [], true);
    const json = JSON.stringify(analyzer.counter.count);

    expect(json).toMatchSnapshot();
  });
});
