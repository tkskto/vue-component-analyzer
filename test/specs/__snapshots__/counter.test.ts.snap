// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`counter's count test counterTest.vue 1`] = `"{\\"/test/fixture/counterTest/counterTest.vue\\":1,\\"/test/components/nested2.vue\\":1,\\"/test/components/child.vue\\":2,\\"/test/components/child2.vue\\":2,\\"/test/fixture/counterTest/counterComponent.vue\\":1,\\"/test/fixture/counterTest/counterComponent2.vue\\":1}"`;
